### PR TITLE
Hack: Add pinned app icon to the desktop app grid to launch the experience

### PR DESCRIPTION
**Pull request by erikos**
DBus Interfaces: com.endlessm.HackableApp
- Method: PulseFlipToHackButton, being able to start/stop pulsating the button from the quest
- Property: ToolboxVisible, property used in the fading animation when transitioning to the desktop after the Unlock puzzle is solved

DBus Interfaces: org.gnome.Shell
- Method: Add MinimizeAll method in order to land on the desktop after the puzzle is solved

Resources:
- add a new flip-to-hack-button png for the pulsate effect
- add the two icons Hack mode on/off for the pinned app

GSettings:
- remove the "enable-code-view" GSetting
- add the "hack-mode-enabled" boolean: this controls if the FtH button is available and Hack mode artefacts like cursor and desktop background

Clubhouse:
- remove the launcher button (right screen side) from the Shell code
- adapt to new packaging of the Clubhouse (Katamari)
- add HackAppIcon which extends ViewIcon and will change depending on the current Hack mode
- remove DiscoveryFeedButton if Clubhouse is installed

https://phabricator.endlessm.com/T27420

### CODE REVIEW

File: js/ui/appDisplay.js [hunk 1]
```diff
@@ -667,11 +676,25 @@ var AllView = class AllView extends BaseAppView {
         };
     }
 
+    _ensureHackAppIcon() {
+        if (this._hackAppIcon)
+            return;
+
+        this._hackAppIcon = new HackAppIcon(this);
+        this._hackAppItem = {
+            get_name: () => { return this._hackAppIcon.getName(); }
```
Comment (ptomato) [inline js/ui/appDisplay.js]: I assume the arrow function is for intentionally capturing `this`? I would advise against doing that, it's confusing and easy to misunderstand. If you have to do it, clarify it with a comment.

File: js/ui/appDisplay.js [hunk 2]
```diff
@@ -2743,3 +2780,80 @@ class AppCenterIcon extends AppIcon {
         return true;
     }
 });
+
+const HackAppIconState = {
+    ACTIVATED: ViewIconState.NUM_STATES + 2,
```
Comment (ptomato) [inline js/ui/appDisplay.js]: Where are these values coming from?

File: js/ui/appDisplay.js [hunk 3]
```diff
@@ -2743,3 +2780,80 @@ class AppCenterIcon extends AppIcon {
...
+
+    _createIcon(iconSize) {
+        let iconUri = 'resource:///org/gnome/shell/theme/hack-button-on.png';
+        if (this.iconState === HackAppIconState.DEACTIVATED)
+            iconUri = 'resource:///org/gnome/shell/theme/hack-button-off.png';
+
+        let iconFile = Gio.File.new_for_uri(iconUri);
+        let gicon = new Gio.FileIcon({ file: iconFile });
+
+        return new St.Icon({ gicon: gicon,
```
Comment (ptomato) [inline js/ui/appDisplay.js]: minor: can be shortened, `return new St.Icon({gicon, iconSize});`

File: js/ui/components/clubhouse.js [hunk 4]
```diff
@@ -830,24 +702,37 @@ var Component = GObject.registerClass({
...
+        if (this.proxy)
+            return this.proxy;
+
+        const clubhouseInstalled = !!getClubhouseApp();
+        if (!clubhouseInstalled) {
+            log('Cannot construct Clubhouse proxy because Clubhouse app was not found.');
+        } else {
+            try {
+                this.proxy = new Gio.DBusProxy({ g_connection: Gio.DBus.session,
+                                            g_interface_name: this._proxyInfo.name,
```
Comment (ptomato) [inline js/ui/components/clubhouse.js]: minor: Alignment is off here.

File: data/org.gnome.shell.gschema.xml.in [hunk 5]
```diff
@@ -147,6 +147,11 @@
         their status icons (e.g. Steam, Dropbox, Skype...).
       </description>
     </key>
+    <key type="b" name="hack-mode-enabled">
```
[the remark this answers is not in the document]
Comment (danigm) [reply data/org.gnome.shell.gschema.xml.in]: Yes, are equivalent, but the hack-mode-enabled will be used to enable/disable the whole hack experience. We started using that key, but the name is a bit confused so we changed to the new one.

File: js/ui/appDisplay.js (hunk 1 above)
Comment (danigm) [reply js/ui/appDisplay.js]: This is a copy of the same code for the `AppCenterIcon` that's just above.

What do you suggest as a replacement of this? The first thing that comes to my mind is to assign directly to the `getName` method, something like this:

```
this._hackAppItem = { get_name: this._hackAppIcon.getName };
```

File: js/ui/appDisplay.js (hunk 2 above)
Comment (danigm) [reply js/ui/appDisplay.js]: This comes from: https://github.com/endlessm/gnome-shell/blob/master/js/ui/appDisplay.js#L2660

The `AppCenterIconState` introduces two new states, and what I've done here is to add two more. I can add this as a comment in the code to clarify why +2 and +3.

File: js/ui/appDisplay.js (hunk 3 above)
Comment (danigm) [reply js/ui/appDisplay.js]: the `iconSize` is not the key for the dict for `St.Icon`, the key is `icon_size`.

Comment (ptomato) [reply js/ui/appDisplay.js]: Oh, I misread that. But by coincidence, my comment could still apply :smile: Both `icon_size` and `iconSize` are accepted for GObject property names, but I see that if you're using underscores consistently then using the shorthand here would be confusing.

File: js/ui/appDisplay.js (hunk 1 above)
Comment (ptomato) [reply js/ui/appDisplay.js]: I would use `bind` which makes it more obvious that the `this` is not what you think when you are calling `get_name`:

```js
get_name: this._hackAppIcon.getName.bind(this),
```

File: data/dbus-interfaces/com.endlessm.HackableApp.xml [hunk 6]
```diff
@@ -1,5 +1,8 @@
 <node>
   <interface name="com.endlessm.HackableApp">
+    <method name="PulseFlipToHackButton">
```
[the remark this answers is not in the document]
Comment (danigm) [reply data/dbus-interfaces/com.endlessm.HackableApp.xml]: done

File: js/ui/shellDBus.js [hunk 7]
```diff
@@ -131,6 +131,12 @@ var GnomeShell = class {
         Main.overview.viewSelector.appDisplay.selectApp(id);
     }
 
+    MinimizeAll() {
+        global.get_window_actors().filter(actor => {
```
[the remark this answers is not in the document]
Comment (danigm) [reply js/ui/shellDBus.js]: done

File: js/ui/appDisplay.js [hunk 8]
```diff
@@ -93,11 +93,7 @@ const EOS_REPLACED_BY_KEY = 'X-Endless-Replaced-By';
 const EOS_NEW_ICON_ANIMATION_TIME = 0.5;
 const EOS_NEW_ICON_ANIMATION_DELAY = 0.7;
 
-const {CLUBHOUSE_ID} = imports.ui.components.clubhouse;
-
-function _getClubhouse() {
-    return Shell.AppSystem.get_default().lookup_app(CLUBHOUSE_ID + '.desktop');
-}
+const Clubhouse = imports.ui.components.clubhouse;
```
[the remark this answers is not in the document]
Comment (danigm) [reply js/ui/appDisplay.js]: done

File: js/ui/appDisplay.js (hunk 1 above)
Comment (danigm) [reply js/ui/appDisplay.js]: done

File: js/ui/components/clubhouse.js (hunk 4 above)
Comment (danigm) [reply js/ui/components/clubhouse.js]: fixed

File: js/ui/appDisplay.js (hunk 2 above)
Comment (ptomato) [reply js/ui/appDisplay.js]: It seems a bit fragile to add values to an existing enum all over the place, what if `AppCenterIconState` gets another value added? But I guess if we do it already for the app center, it's out of scope to fix it for Hack.

To minimize the risk, how about renaming `AppCenterIconState`  to `DownstreamEndlessIconState` and merging these values into there?

Comment (danigm) [reply js/ui/appDisplay.js]: Done, I've renamed the `AppCenterIconState` to `DownstreamEndlessIconState` and merged there the AppCenter and the HackIcon keys.

File: js/ui/codeView.js [hunk 9]
```diff
@@ -1470,11 +1474,14 @@ var CodeViewManager = GObject.registerClass({
...
 
+        let gtkId = actor.meta_window.get_gtk_application_id();
+
         // The custom X-Endless-Hackable key has the last word always
         if (appInfo.has_key(_HACKABLE_DESKTOP_KEY)) {
             if (!appInfo.get_boolean(_HACKABLE_DESKTOP_KEY))
                 return false;
-        } else {
+        // The hack unlock has the com.endlessm.Clubhouse id in the katamari flatpak
+        } else if (gtkId !== 'com.endlessm.HackUnlock' && gtkId !== 'com.endlessm.HackToolbox') {
```
Comment (ptomato) [inline js/ui/codeView.js]: I think the string here is correct because you are checking the app ID, not the namespace of the flatpak, but I am a bit confused by the comment.

File: js/ui/codeView.js [hunk 10]
```diff
@@ -61,6 +61,10 @@ function _ensureAfterFirstFrame(win, callback) {
 
 function _getAppId(win) {
     let app = Shell.WindowTracker.get_default().get_window_app(win);
+    let gtkId = win.get_gtk_application_id();
```
Comment (ptomato) [inline js/ui/codeView.js]: Should this change in `_getAppId()` be split into a separate commit to go upstream? Is it a general fix for flatpaks that ship more than one app?

File: js/ui/discoveryFeedButton.js [hunk 11]
```diff
@@ -4,6 +4,7 @@ const { Clutter, Gio, GLib, GObject, St } = imports.gi;
 
 const Lang = imports.lang;
 const Main = imports.ui.main;
+const Clubhouse = imports.ui.components.clubhouse;
```
Comment (ptomato) [inline js/ui/discoveryFeedButton.js]: I understood from the Slack discsussion that this commit isn't needed anymore?

File: js/ui/codeView.js (hunk 10 above)
Comment (danigm) [reply js/ui/codeView.js]: I don't know if this will be interesting for upstream. This does the work for gtk apps that are inside the same flatpak with different desktop files, so the desktop file is not exported. This is a rare case that shouldn't happen in usual apps. We've this problem because we've the `com.endlessm.HackUnlock` and `com.endlessm.HackToolbox` inside the katamari and we won't export the desktop files, but if a flatpak has several apps it usually uses the same prefix and exports all the .desktop files, like libreoffice does.

So I think this is something to work with our especial case flatpak deployment.

File: js/ui/discoveryFeedButton.js (hunk 11 above)
Comment (danigm) [reply js/ui/discoveryFeedButton.js]: Yes, I'll remove this commit to don't touch the discovery feed.

File: js/ui/codeView.js (hunk 9 above)
Comment (danigm) [reply js/ui/codeView.js]: I've tried to explain better in the comment:

```
        // HackUnlock and HackToolbox are inside the com.endlessm.Clubhouse flatpak
        // and have the Clubhouse appInfo, so we should ignore those cases here
        // to be able to show the HackUnlock and HackToolbox windows
```

**Comment by erikos** [comment]
> Thanks for the rebase comments, they will be helpful. A few minor comments left.
> 
> Is there an easy way to test this by downloading a Hack image or something like that? Or do I have to build/install the shell and Katamari?

The best way is to build the shell with this branch (e.g. eos-dev-unlock) and use Katamari to install the minimal Hack version. Specify the hack2 branch for the Clubhouse in the config file.

**Comment by danigm** [comment]
We've a custom OBS gnome-shell repository, I've just updated that so you can try to use it if you don't want to build the shell. Here you can find instructions to install the shell package and the katamari flatpaks:

https://phabricator.endlessm.com/w/howto/hack2-gnome-shell/

File: js/ui/appDisplay.js [hunk 12]
```diff
@@ -620,12 +623,17 @@ var AllView = class AllView extends BaseAppView {
...
+        this._maybeAddHackIcon();
         super.loadGrid();
     }
 
     getLayoutIds() {
         let layoutIds = super.getLayoutIds();
 
+        // Only show the hack icon if the clubhouse app is in the system
+        if (Clubhouse.getClubhouseApp())
+            layoutIds.splice(0, 0, HACK_APP_ID);
```
Comment (ptomato) [inline js/ui/appDisplay.js]: nitpick: Maybe use `layout.unshift(HACK_APP_ID)` here, `splice()` tends to be confusing

**Comment by erikos** [comment]
As discussed on Slack: what you have seen with the non-minified apps is because while the Unlock app was already open some of the apps have been revealed (e.g. tabbed to it). The current code (https://github.com/endlessm/hack-toy-apps/commit/aa145be7ccfbde2394e8268c3251ef93c1eacc9b)  in the Unlock app does minimize the apps on application start.

This is a bit fragile. Minimizing the apps should be done at the quest level before landing on the desktop, same as we do when changing the cursor to the Hack one.

But as you said, this is independent from the shell code.

File: js/ui/appDisplay.js (hunk 12 above)
Comment (danigm) [reply js/ui/appDisplay.js]: I've used unshift as you suggested.
